src/bloom/hierarchical.rs

```diff
@@ -0,0 +1,335 @@
+use std::iter::FromIterator;
+
+use super::*;
+
+/// Container for types which is support bloom filtering
+#[derive(Debug)]
+pub struct HierarchicalBloom<Child>
+where
+    Child: BloomProvider + Send + Sync,
+{
+    inner: HierarchicalBloomInner,
+    children: Vec<Child>,
+    group_size: usize,
+}
+
+#[derive(Debug)]
+enum HierarchicalBloomInner {
+    /// Inner filters
+    Vec(Option<Bloom>, Vec<HierarchicalBloomInner>),
+    /// Leaf
+    Item(usize),
+}
+
+impl Default for HierarchicalBloomInner {
+    fn default() -> Self {
+        Self::Vec(Default::default(), Default::default())
+    }
+}
+
+impl<Child> Default for HierarchicalBloom<Child>
+where
+    Child: BloomProvider + Send + Sync,
```

Comment:
You don't need `Send` and `Sync` here. There should be as little trait constraints as possible everywhere, as it helps with maintenance